Fixes #4151 - Reports STI

shlomizadok opened the pull request.

shlomizadok committed:
fixes #4151 - enable reports STI

ohadlevy commented on app/models/puppet_report.rb:
```diff
@@ -0,0 +1,20 @@
+class PuppetReport < Report
+  def self.model_name
+    superclass.model_name
```
why is that needed? IMHO you can use include Foreman::STI to get rid of most related issues?

ohadlevy replied:
ah, you use it already, is it needed?

ohadlevy replied:
actually i take it back, please ignore :)

ohadlevy commented on app/models/report.rb:
```diff
@@ -132,14 +135,10 @@ def no_report
   end
 end
+
+require_dependency 'puppet_report'
```
maybe add it to config.autoload_paths under config/application.rb?

ohadlevy commented on db/migrate/20150705131449_add_type_to_reports.rb:
```diff
@@ -0,0 +1,5 @@
+class AddTypeToReports < ActiveRecord::Migration
+  def change
+    add_column :reports, :type, :string, :default => 'PuppetReport'
```
two issues:
1. what about salt/chef reports? they would also be migrated to puppet report?
2. type should be null => false

ohadlevy replied:
cc @ares @stbenjam

stbenjam replied:
I would prefer if there was a generic ConfigReport that has the general idea behind a config management system: system status, resources, results, etc, and allow config management platforms to extend it if necessary.  Does that make sense?

That would let puppet/salt/chef customize it, while still providing the base data to be the source of data for the host status, dashboard, etc.

ohadlevy commented on app/models/report.rb:
```diff
@@ -132,14 +135,10 @@ def no_report
   end
 end
+
+require_dependency 'config_report'
```
I don't think this should be needed. autoloader should handle it afaik

shlomizadok replied:
I must say that I cowardly copied from https://github.com/theforeman/foreman/blob/develop/app/models/template.rb#L68

ares replied:
It is required for dev env, if you call `Report.all` in dev env before ConfigReport was loaded the STI does not work correctly.

ohadlevy commented on db/migrate/20150705131449_add_type_to_reports.rb:
```diff
@@ -0,0 +1,5 @@
+class AddTypeToReports < ActiveRecord::Migration
+  def change
+    add_column :reports, :type, :string, :default => 'ConfigReport'
```
null => false?

shlomizadok replied:
is `null => false` needed if `:default => 'ConfigReport'` ?

domcleal commented on app/services/report_importer.rb:
```diff
@@ -2,14 +2,39 @@ class ReportImporter
+  # to be defined in children
+  def report_name_class
+    # to raise NotImplementedError ?
+    Foreman::Deprecation.deprecation_warning(1.12, "This shit is deprecated, wtf are you doing?")
```
Not at all appropriate, please remove it and fix the question above.

domcleal commented on app/models/config_report.rb:
```diff
@@ -0,0 +1,20 @@
+    ConfigReportImporter.import(report, proxy_id)
+  end
+
+  def summaryStatus
```
I'd suggest changing this to underscores from camel case if you're moving it.

domcleal commented:
Please address any questions/TODOs in the code, there are a few.

shlomizadok commented:
Addressed the questions and changed (:cold_sweat: hope all is well)

shlomizadok commented:
Added @ohadlevy comments too

tstrachota commented on app/views/api/v1/reports/show.json.rabl:
```diff
@@ -13,5 +13,5 @@ child :logs do
 
 node :summary do |report|
-	report.summaryStatus
+	report.summary_status
```
Wrong indentation, there's tab instead of 2 spaces.

ares commented on app/models/report.rb:
```diff
@@ -82,7 +83,9 @@ def runtime
   def self.import(report, proxy_id = nil)
-    ReportImporter.import(report, proxy_id)
+    # to raise NotImplementedError ?
+    Foreman::Deprecation.deprecation_warning(1.12, "Report model has turned to be STI, please use child classes")
```
I think the version should be string

shlomizadok replied:
Works as an integer too :)

ares commented on app/models/config_report.rb:
```diff
@@ -0,0 +1,20 @@
+  def self.report_status
+    "status"
+  end
+end
```
missing new line at the end of file

ares commented on app/services/config_report_importer.rb:
```diff
@@ -0,0 +1,9 @@
+class ConfigReportImporter < ReportImporter
+  def self.authorized_smart_proxy_features
+    ['Puppet', 'Chef Proxy']
```
I think don't need `'Chef Proxy'`, since it's called `'Chef'` and it's defined by chef plugins. It should be dropped from seeds as well (saying it's out of scope for this PR is OK)

EDIT: saying that I think we need API to add more features here, e.g. Chef or Salt

shlomizadok replied:
Original importer had Puppet and Chef feature.
Other importers (features) can inherit from `ReportImporter` and add their own feature?

ares commented on app/services/report_importer.rb:
```diff
@@ -2,14 +2,38 @@ class ReportImporter
+
+  def self.authorized_smart_proxy_features
+    Rails.logger.debug("Importer #{self} does not implement authorized_smart_proxy_features.")
+    Foreman::Deprecation.deprecation_warning(1.12, "ReportImporter is a superclass, please use child classes. Maybe try: ConfigReportImporter?")
```
1.12 should be string I think

shlomizadok replied:
ditto, integer seems ok

domcleal replied:
That's a float, not an int, so extra reason to avoid it.  The intention was for it to be a string, but the class doesn't enforce it, thought it probably would.

ares replied:
yup, float was my concern

ares commented on app/services/report_importer.rb:
```diff
@@ -2,14 +2,38 @@ class ReportImporter
-    ['Puppet', 'Chef Proxy']
+  # to be defined in children
+  def report_name_class
+    Foreman::Deprecation.deprecation_warning(1.12, "ReportImport is a supperclass, please use child classes to define 'report_class_name'")
```
1.12 should be string I think, also I'm wondering if any plugin uses this class directly, maybe we don't need to keep backwards compatibility here

ares commented on app/models/report.rb:
```diff
@@ -132,13 +135,6 @@ def no_report
     false
   end
 
-  def summaryStatus
```
just thank you for renaming!

shlomizadok commented:
Stringified deprecation warning version number. Changed "Chef Proxy" to "Chef" + migration to existing installations. (Please note there was a duplication with https://github.com/theforeman/foreman_chef/blob/master/db/seeds.rb#L1)

shlomizadok commented:
re[test] due to un-related Katello failure

ares commented on app/services/config_report_importer.rb:
```diff
@@ -0,0 +1,9 @@
+  def report_name_class
+    ConfigReport
+  end
+end
```
missing new line at the end of file

shlomizadok replied:
Fixed. Thanks :)

2015-07-15 14:31 GMT+03:00 Marek Hulán notifications@github.com:

> In app/services/config_report_importer.rb
> https://github.com/theforeman/foreman/pull/2530#discussion_r34669168:
> 
> > @@ -0,0 +1,9 @@
> > +class ConfigReportImporter < ReportImporter
> > -  def self.authorized_smart_proxy_features
> > -    ['Puppet', 'Chef']
> > -  end
> >   +
> > -  def report_name_class
> > -    ConfigReport
> > -  end
> >   +end
> 
> missing new line at the end of file
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/theforeman/foreman/pull/2530/files#r34669168.

domcleal commented on app/services/report_importer.rb:
```diff
@@ -2,14 +2,38 @@ class ReportImporter
+
+  def self.authorized_smart_proxy_features
+    Rails.logger.debug("Importer #{self} does not implement authorized_smart_proxy_features.")
+    Foreman::Deprecation.deprecation_warning('1.12', "ReportImporter is a superclass, please use child classes. Maybe try: ConfigReportImporter?")
```
I don't understand the deprecation warnings here - they're going to fire all the time with the ReportsController and I don't see what replaces them.  The controller is always going to call ReportImporter.authorized_smart_proxy_features, not some Report or ReportImporter subclass.

They're also for a new method, which doesn't make a lot of sense.

domcleal replied:
Hm, possibly ignore this, I think the issue might be in the controller.

domcleal commented on app/controllers/api/v2/reports_controller.rb:
```diff
@@ -7,7 +7,7 @@ class ReportsController < V2::BaseController
       before_filter :setup_search_options, :only => [:index, :last]
 
-      add_smart_proxy_filters :create, :features => ReportImporter.report_features
+      add_smart_proxy_filters :create, :features => ReportImporter.authorized_smart_proxy_features
```
Oh, regarding my comment about deprecation of authorized_smart_proxy_features, did you mean to actually keep `ReportImporter.report_features` here which calls that method on each of the known report importers?

domcleal replied:
Still an issue, or use ConfigReportImporter specifically?

domcleal commented on db/migrate/20150714151223_rename_chef_proxy.rb:
```diff
@@ -0,0 +1,9 @@
+class RenameChefProxy < ActiveRecord::Migration
+  def up
+    chef_feature = Feature.find_by_name("Chef Proxy")
+    chef_feature.update_attribute(:name, "Chef") if chef_feature
```
If foreman_chef's loaded, this might lead to a duplicate - should we delete Chef Proxy if it's there, leaving the plugin to define Chef?

domcleal commented on db/migrate/20150714151223_rename_chef_proxy.rb:
```diff
@@ -0,0 +1,9 @@
+  end
+
+  def down
+  end
```
todo?

domcleal commented on db/seeds.d/11-smart_proxy_features.rb:
```diff
@@ -1,5 +1,5 @@
 # Proxy features
-[ "TFTP", "DNS", "DHCP", "Puppet", "Puppet CA", "BMC", "Chef Proxy", "Realm", "Facts" ].each do |input|
+[ "TFTP", "DNS", "DHCP", "Puppet", "Puppet CA", "BMC", "Chef", "Realm", "Facts" ].each do |input|
```
Time to remove this altogether, see above?

domcleal commented on app/models/report.rb:
```diff
@@ -82,7 +83,9 @@ def runtime
   end
 
   def self.import(report, proxy_id = nil)
```
This method's still called from our regular API controller isn't it?

domcleal replied:
Looks like the controller needs refreshing to use this capability, and so it selects the right importer.

domcleal commented on app/models/report.rb:
```diff
@@ -82,7 +83,9 @@ def runtime
 
   def self.import(report, proxy_id = nil)
-    ReportImporter.import(report, proxy_id)
+    # to raise NotImplementedError ?
```
Probably remove this?

domcleal commented on app/services/config_report_importer.rb:
```diff
@@ -0,0 +1,9 @@
+class ConfigReportImporter < ReportImporter
+  def self.authorized_smart_proxy_features
+    ['Puppet', 'Chef']
```
I'd suggest moving this to foreman_chef as part of 1.10 compatibility, probably by overriding it?

Would foreman_chef & salt implement their own types of the reports?  It appears not, so I'd guess they'd alias chain it or we'd need to introduce another API.

domcleal replied:
Still?  Remove this and move to the plugin possibly?

domcleal commented:
Should the reports controllers be replaced by ones specifically for config reports, so we have something more analogous to the templates STI migration?

At the moment it's more like facts, which is OK as they're all a more simple key/value structure and can be reliably displayed in the same way.

It strikes me that reports could vary in type in more ways, so having a single controller, single set of views etc is going to be very limiting.

domcleal commented:
Separate routes would also help you implement this part of the feature request:
- As a user I want to have be able to set different report permissions per report types. (implies report STI)

shlomizadok commented:
@domcleal thanks for the comments and insights. I have added separate routes and controllers.

shlomizadok commented:
[test]

ohadlevy commented on app/controllers/reports_controller.rb:
```diff
@@ -2,31 +2,23 @@ class ReportsController < ApplicationController
   include Foreman::Controller::AutoCompleteSearch
 
   before_filter :setup_search_options, :only => :index
+  before_filter :deprecated
```
why do we need to deprecate a ui controller? can't we fix it in apache/routes instead of keeping a controller to redirect?

domcleal replied:
Indeed, just remove this controller as we did with ConfigTemplates.  Also remove its old views.

shlomizadok commented:
@domcleal - I have removed the reports_controller and added a `DeprecationRedirect` class, so it may be used by other deprecated routes. Let me know if it makes sense. Thanks.

shlomizadok commented:
[test]

shlomizadok commented:
Nothing like a good ole' rebase to make tests pass :)

domcleal commented on app/services/report_importer.rb:
```diff
@@ -2,14 +2,38 @@ class ReportImporter
+    importers[key.to_sym] = klass
+  end
+
+  def self.fact_features
```
Unused, and not named correctly anyway!

domcleal commented on app/services/report_importer.rb:
```diff
@@ -2,14 +2,38 @@ class ReportImporter
   delegate :logger, :to => :Rails
   attr_reader :report
 
+  def self.importer_for(type)
```
The registration of a type to an importer is unused in the current arrangement, as you've got a controller per type, whereas facts (which this is copied from) had a single controller with the type passed in.

domcleal commented on app/controllers/api/v1/reports_controller.rb:
```diff
@@ -40,6 +41,12 @@ def last
         @report = Report.authorized(:view_reports).includes(:logs => [:message, :source]).find(max_id)
```
Should this controller use ConfigReport as its base class?

domcleal commented on test/factories/reports_related.rb:
```diff
@@ -6,6 +6,10 @@
   end
 
+  factory :config_report, :parent => :report, :class => 'ConfigReport' do
+    # Add specific Puppet report attrs if needed
```
Just remove the block, not needed.

domcleal commented:
1) I'm unsure the redirect in the UI is necessary, we didn't do it with templates.  I suppose the only possibility is that people who've bookmarked the page don't have the links broken, not sure if it's worthwhile or not?

2) Bookmarks for the "reports" controller need updating to point to the right controller, I think.

3) Should permissions be renamed to config_reports?

domcleal commented:
4)  "As a user I want to search hosts based on report type and its specific status." - how will this be done, do we need ways to combine report statuses or for now only report on ConfigReports from the dashboard, status icons etc?

shlomizadok commented:
@domcleal - Thanks for you comments, they are really helpful!
some answers:
1) I thought that "/reports" is more highly profiled and may be bookmarked / directly accessed so at least in the next version to use that redirect. I can remove it though.

2) How do I do that? migration?

3) Don't think so. 

4) Good question, I think for the ease of use, let's for now use only ConfigReports from dashboard / status icons - as it is similar to current status

I will apply all code related comments.
Let me know what you think.
Thanks a lot :+1:

domcleal commented:
1) Ok, fair enough.

2) Yes, just change the controller name (which is a separate column) in the bookmarks table.

3) I'm inclined to say it should be renamed as templates were.

domcleal commented on lib/deprecation_redirect.rb:
```diff
@@ -0,0 +1,13 @@
+class DeprecationRedirect
```
Perhaps put this in app/, er services(?) so it can be auto-reloaded by Rails.

domcleal commented on lib/deprecation_redirect.rb:
```diff
@@ -0,0 +1,13 @@
+class DeprecationRedirect
+  attr_reader :redirect, :params
+  def initialize(redirect, params = nil, version, message)
+    Foreman::Deprecation.deprecation_warning("#{version}", "#{message}")
```
Remove string interpolation, just pass version, message directly.

domcleal commented on config/routes.rb:
```diff
@@ -10,7 +10,12 @@
+  # Redirect reports with deprecation warning
+  get "/reports", :to => redirect { |params, request| DeprecationRedirect.new("config_reports", params, '1.12', 'Reports are renamed to Config Reports').redirect_url }
+  get "/reports/:id", :to => redirect { |params, request| DeprecationRedirect.new("config_reports", params, '1.12', 'Reports are renamed to Config Reports').redirect_url }
+  delete "/reports/:id", :to => redirect { |params, request| DeprecationRedirect.new("config_reports", params, '1.12', 'Reports are renamed to Config Reports').redirect_url }
```
I wouldn't bother redirecting a delete.

domcleal commented on lib/deprecation_redirect.rb:
```diff
@@ -0,0 +1,13 @@
+  end
+
+  def redirect_url
+    return "/#{@redirect}/#{@params[:id]}" if @params && @params[:id]
```
I haven't tested this feature yet, but should all params be passed through, so things like search work?

shlomizadok commented:
[test] on random (?) katello failure

stbenjam commented:
This PR makes me so happy. :grin: 

> @domcleal: It strikes me that reports could vary in type in more ways, so having a single controller, single set of views etc is going to be very limiting.

A single controller should be fine if the report object is being delivered to the view, no? The views though is something I'd want to customize, what if I want to add some things to a Salt report? Like another pie chart or some tables.

domcleal commented:
> A single controller should be fine if the report object is being delivered to the view, no? The views though is something I'd want to customize, what if I want to add some things to a Salt report? Like another pie chart or some tables.

For something that isn't a ConfigReport, e.g. a `FooReport < Report` then the index probably needs to be different to show different columns (number of service restarts is probably not interesting) and could have different permissions, which suits having a separate FooReportsController.  This is the case that I was describing above and seems to be in the ticket (e.g. ABRT).

Salt is probably a different case.. I'd guess you'd actually want to specialise ConfigReport by subclassing that instead?  So you'd have mostly the same functionality and could share 95% of what ConfigReport already has.  In theory the same controller could work, but it probably would need some further changes, e.g. to search for not only ConfigReport, but subclasses such as SaltConfigReport, and then a way to load a SaltConfigReport-specific view from ConfigReportsController.

Alternatively as it stands you'd add `SaltConfigReport < Report` and probably have to reimplement the views, controllers etc.  I think we ought to enable the second specialising style, but probably as a second patch.  Make sense?

stbenjam commented:
> In theory the same controller could work, but it probably would need some further changes, e.g. to search for not only ConfigReport, but subclasses such as SaltConfigReport, and then a way to load a SaltConfigReport-specific view from ConfigReportsController.

Yea, that's kind of what I was hoping for out of the box to begin with.  I'd like to have the SaltConfigReport to be able to tweak it a little, but still have it be the data source for the host and dashboard statuses and show up under the listing for all ConfigReports.

I guess my reasoning is, most (all?) config platforms have the concept of system is OK, made changes, or bad state, and also have a concept of a set of resources that were applied, so it should all be lumped together.

I realize other reports have different needs, like ABRT, so it makes sense they would go the other way, but for config management systems, the other way makes me have to maintain a separate but nearly identical set of code just for Salt reports, including my own dashboard widgets, etc, right? And lose the ability to affect the global host status in the All Hosts list, for example.

domcleal commented:
> I realize other reports have different needs, like ABRT, so it makes sense they would go the other way, but for config management systems, the other way makes me have to maintain a separate but nearly identical set of code just for Salt reports, including my own dashboard widgets, etc, right? And lose the ability to affect the global host status in the All Hosts list, for example.

Yeah exactly, I didn't mean to recommend doing that.  I agree we should support subclassing of ConfigReport, but perhaps we follow up with it (if @shlomizadok or somebody is willing), so as not to conflate the two styles of STI extension.  One's hard enough to review!

shlomizadok commented:
@domcleal, I will follow up on that. ( @stbenjam, let's propose this issue together on redmine - and I will pick it from there )

domcleal commented on app/services/report_importer.rb:
```diff
@@ -2,14 +2,22 @@ class ReportImporter
-    ['Puppet', 'Chef Proxy']
+  # to be defined in children
+  def report_name_class
+    Foreman::Deprecation.deprecation_warning('1.12', "ReportImport is a supperclass, please use child classes to define 'report_class_name'")
```
Is a deprecation warning required here?  I'm trying to see how it might be called, and I think it's only by defining a new STI type anyway, which means you're already using this new feature.

domcleal commented on app/services/config_report_importer.rb:
```diff
@@ -0,0 +1,9 @@
+class ConfigReportImporter < ReportImporter
```
This class should take on some of the code from ReportImporter I think that is Puppet-specific, e.g. setting puppet_status, email notifications.

shlomizadok replied:
@domcleal Should all the import method move here? it seems so

domcleal replied:
I'd say most of it should.. things like timings and log message storage helpers would be useful in the base class I'd say.

domcleal commented on app/services/foreman/access_permissions.rb:
```diff
@@ -701,15 +701,18 @@
-                                      :"api/v2/reports" => [:destroy]
-    }
-    map.permission :upload_reports,  {:"api/v2/reports" => [:create] }
+    map.permission :view_reports, {:"api/v1/reports"        => [:index, :show, :last],
```
I think we should still consider renaming permissions too?

ohadlevy commented:
please dont forget to update the reports rake tasks (expire, mails etc)

domcleal commented:
In what way?  Split up expiration depending on the STI subclass rather than expiring all reports?

ohadlevy commented:
@domcleal yeah, different type of reports can have different expiry policy?

shlomizadok commented:
@domcleal renamed permissions, modified the ReportImporter and ConfigReportImporter to include puppet specific stuff.

@ohadlevy, not sure we need to change the rake task. Do you mean that it should change to explicitly ConfigReport rake task and STI subclasses may create their own tasks?

shlomizadok commented:
Fixed permissions: Added migration and changed in 03-permissions

ohadlevy commented:
@shlomizadok I mean this line https://github.com/shlomizadok/foreman/blob/7f6aecd23a3e67db79f4b04bb46208c802f3eef3/lib/tasks/reports.rake#L21. AFAIK it should only be relevant to ConfigReport and not Report?

shlomizadok commented:
@ohadlevy, so I'm not sure. Maybe I want to expire all reports (ConfigReport, ArfReport, SaltReport) with that task?

ohadlevy commented:
@shlomizadok by default, users already have this tasks, imho we dont want
to expire scap reports without thinking about it first :)

shlomizadok commented:
@ohadlevy - changed. [test]

domcleal commented on lib/tasks/reports.rake:
```diff
@@ -18,7 +18,7 @@ namespace :reports do
     conditions[:status] = ENV['status'].to_i if ENV['status']
 
-    Report.expire(conditions)
+    ConfigReport.expire(conditions)
```
Perhaps make this an argument so the same task can be re-used for different subclasses.

domcleal commented on app/controllers/api/v2/reports_controller.rb:
```diff
@@ -59,6 +59,12 @@ def last
         @report = Report.authorized(:view_reports).includes(:logs => [:message, :source]).find(max_id)
```
Lots of references to Report in this controller to change to ConfigReport, and probably override resource_class.

domcleal commented on app/services/report_importer.rb:
```diff
@@ -25,19 +32,13 @@ def import
 
     if host.new_record? && !Setting[:create_new_host_when_report_is_uploaded]
```
There's duplication between ReportImporter.import and ConfigReporter.import.  Could we either - split this up into smaller, more extensible methods that subclasses can hook into (so we have one code path, jumping to subclass methods to say, parse Puppet-specifics), or remove this version?

shlomizadok replied:
Didn't fix on this run, will see to it tomorrow

shlomizadok replied:
Refactored. let me know what you think

domcleal commented on app/models/config_report.rb:
```diff
@@ -0,0 +1,20 @@
+  end
+
+  # puppet report status table column name
+  def self.report_status
```
Nitpick, perhaps call this report_status_column so it's more different to ReportImporter.report_status.

domcleal commented on app/controllers/api/v2/config_reports_controller.rb:
```diff
@@ -0,0 +1,68 @@
+        render :show
+      end
+
+      def resource_class
```
Is this needed, the default won't suffice?

domcleal commented:
1. The menu text and heading on /config_reports will need changing to something else, such as "Config Reports" or "Configuration Management Reports".  I'm a little uncomfortable calling it Config Reports, but expanding the words makes it rather unwieldy.
2. Ditto on the host page ("Reports" button), perhaps only in the tooltip.
3. The "Last report" column on host#index uses Report, should it be ConfigReport?
4. Ditto on dashboard widgets.
5. Ditto on host#show widgets.
6. The Puppet summary (HostMailer summary / Report.summarise) uses Report, rather than ConfigReport.

shlomizadok commented:
@domcleal - I have changed all of the comments above

shlomizadok commented:
I hate typos :/

domcleal commented:
Lots of red tests.

shlomizadok commented:
Ain't that sad :weary: ?
Hope that latest fixes them

shlomizadok commented:
@domcleal tests are :green_apple:

domcleal commented on db/migrate/20150721131324_change_bookmark_report_controller.rb:
```diff
@@ -0,0 +1,8 @@
+class ChangeBookmarkReportController < ActiveRecord::Migration
+  def change
+    bookmarks_with_report_controller = Bookmark.where(:controller => 'reports')
+    bookmarks_with_report_controller.each do | bookmark |
```
Use .update_all instead of a loop.

domcleal commented on db/migrate/20150721131324_change_bookmark_report_controller.rb:
```diff
@@ -0,0 +1,8 @@
+class ChangeBookmarkReportController < ActiveRecord::Migration
+  def change
```
This is only applicable to `up`, it needs splitting in two.

domcleal commented on app/services/menu/loader.rb:
```diff
@@ -43,8 +43,8 @@ def self.load
           menu.item :dashboard,         :caption => N_('Dashboard')
-          menu.item :reports,           :caption => N_('Reports'),
-                    :url_hash => {:controller => '/reports', :action => 'index', :search => 'eventful = true'}
+          menu.item :reports,           :caption => N_('Config Mgmt. Reports'),
```
Perhaps add a divider ("Reports") and then list "Config management" below it?  It's still not terribly pretty.

domcleal commented on lib/tasks/reports.rake:
```diff
@@ -3,22 +3,35 @@ desc <<-END_DESC
     rake reports:expire days=7 RAILS_ENV="production" # expires all reports regardless of their status
     rake reports:expire days=1 status=0 RAILS_ENV="production" # expires all non interesting reports after one day
+    rake reports:expire report_type=my_report, days=3 # expires all reports of type MyReport (underscored style) from the last 3 days.
+    rake reports:expire report_type=MyReport, days=3 # expires all reports of type MyReport (class name style) from the last 3 days.
```
No comma between rake environment arguments

domcleal commented on test/unit/report_test.rb:
```diff
@@ -148,4 +148,28 @@ def setup
+
+    test "Report has default type" do
+      report = Report.new
+      assert_equal(report.type, 'ConfigReport')
```
swap arguments here and below, it should be `assert_equal expected, actual`

domcleal commented on app/services/config_report_importer.rb:
```diff
@@ -0,0 +1,41 @@
+  end
+
+  def import
+    @report = super
```
The parent method already seems to set `@report`

domcleal replied:
We've lost a short-circuit that used to be in the original.  If `@report.save` returns false, then the report was returned immediately.

shlomizadok replied:
Ammended the short-circuit, yet still have to set `@report = super`

domcleal replied:
It's set here though, isn't it?  https://github.com/theforeman/foreman/pull/2530/files#diff-5f2522f8937b1076938bcca2ab931cf1R50

shlomizadok replied:
It is set, yet not available unless I set `@report = super` (so it fails on undefined persisted? for nil class... )
Maybe because it is not in the same context?

domcleal replied:
I thought it was shared.  If you use `report` (the attr_reader) rather than `@report`, is it better?

shlomizadok replied:
nope, tried both. 
I find this very strange

domcleal commented on app/services/config_report_importer.rb:
```diff
@@ -0,0 +1,41 @@
+    import_log_messages
+    # Check for errors
+    notify_on_report_error
+    logger.info("Imported report for #{name} in #{(Time.now - @start_time).round(2)} seconds")
```
Perhaps put this in the base class instead.

domcleal commented on app/services/config_report_importer.rb:
```diff
@@ -0,0 +1,41 @@
+  private
+
+  def notify_on_report_error
+    if @report.error?
```
Use the reader provided, remove the `@`

domcleal commented on app/services/config_report_importer.rb:
```diff
@@ -0,0 +1,41 @@
+
+  private
+
+  def notify_on_report_error
```
You might find this is a useful method to put in the base class, perhaps adding an argument to specify the notification name would make it reusable so other impls can seed a mail notification of their own and use it.

domcleal commented on test/unit/config_report_importer_test.rb:
```diff
@@ -1,6 +1,6 @@
 require 'test_helper'
 
-class ReportImporterTest < ActiveSupport::TestCase
+class ConfigReportImporterTest < ActiveSupport::TestCase
```
Could you try splitting this file up to test the generic ReportImporter and the ConfigReportImporter separately?  Only test the Puppet-specifics in the latter, and see if you can test the ReportImporter in a more generic, non-Puppet way.

shlomizadok commented:
@domcleal - addressed your comments.
Please note that in `ConfigReportImporter.import`, I have to set `@report = super` so haven't changed that.

domcleal commented on app/services/report_importer.rb:
```diff
@@ -95,7 +93,11 @@ def import_log_messages
+    ReportStatusCalculator.new(:counters => raw['status']).calculate
+  end
+
+  def notify_on_report_error(mail_error_state = :puppet_error_state)
```
Make this a mandatory argument?

shlomizadok replied:
I wasn't sure if this is used elsewhere (e.g., plugins), though I think it is not.
Will move to a mandatory argument

domcleal commented on app/services/report_importer.rb:
```diff
@@ -46,13 +47,10 @@ def import
-    inspect_report
+    @report = report_name_class.new(:host => host, :reported_at => time, :status => status, :metrics => raw['metrics'])
+    return report unless @report.save
     logger.info("Imported report for #{name} in #{(Time.now - start_time).round(2)} seconds")
```
I'm not entirely sure what to suggest, but we've lost the timing information around the import_log_messages method.

shlomizadok replied:
right, which is why I thought initially to move it to config_report_importer

domcleal replied:
Perhaps just put a second timer inside the log messages method.

domcleal replied:
Or move it up the stack, into a wrapping method?

shlomizadok replied:
@domcleal not sure I follow on "Or move it up the stack..."

domcleal replied:
As in, have `ReportImporter#import`, which looks like:

```
def import
  start_time = Time.now
  import_internal
  logger.info("Imported report...")
end
```

It calls `import_internal`, which is where the guts of the method go, including in ConfigReportImporter.  (import_internal is a terrible name, please don't copy that idea.)

shlomizadok replied:
This actually makes more sense.
btw, haven't thought if copying the idea until you mentioned it. :smile_cat:

shlomizadok commented:
Not sure the name I came up with is better.
Changed (makes much more sense).

domcleal commented on app/services/config_report_importer.rb:
```diff
@@ -0,0 +1,21 @@
+  private
+
+  def create_report_and_logs
+    @report = super
```
I still can't see a use for setting this instance variable or the attr_reader above:

<pre>
From: /home/dcleal/code/foreman/foreman/app/services/config_report_importer.rb @ line 15 ConfigReportImporter#create_report_and_logs:

    13: def create_report_and_logs
    14:   super
 => 15:   binding.pry
    16:   return report unless report.persisted?
    17:   # Store all Puppet message logs
    18:   import_log_messages
    19:   # Check for errors
    20:   notify_on_report_error(:puppet_error_state)
    21: end

[1] pry(#&lt;ConfigReportImporter>)> 
[2] pry(#&lt;ConfigReportImporter>)> report
=> #&lt;ConfigReport id: 1345, host_id: 109, reported_at: "2015-08-24 09:01:02", created_at: "2015-08-24 09:01:05", updated_at: "2015-08-24 09:01:05", status: 1, metrics: "--- !ruby/hash:ActiveSupport::HashWithIndifferentAc...", type: "ConfigReport">
</pre>

shlomizadok replied:
@domcleal - this is weird.
It works on foreman as expected, yet tests fail with:

``` ruby
test_0003_Report has child classes(ReportTest::Report STI):
NoMethodError: undefined method `persisted?' for nil:NilClass
    /home/szadok/dev/upstream/foreman/app/services/config_report_importer.rb:17:in `create_report_and_logs'
```

Any clue how to tackle this? (thanks)

domcleal replied:
I can't replicate the failure.

domcleal commented on app/services/menu/loader.rb:
```diff
@@ -43,12 +43,14 @@ def self.load
           menu.item :trends,            :caption => N_('Trends')
           menu.item :audits,            :caption => N_('Audits')
+          menu.divider                  :caption => N_('Reports')
+          menu.item :reports,           :caption => N_('Config Mgmt. Reports'),
```
Leave out "Reports" now you have a section heading, like the Configure menu.  Perhaps see if expanding "Management" fits sensibly now?

domcleal replied:
Or just expand it anyway, since we should avoid abbreviations.

domcleal commented on app/services/report_importer.rb:
```diff
@@ -108,10 +90,33 @@ def inspect_report
+    status = report_status
+
+    # we update our host record, so we won't need to lookup the report information just to display the host list / info
+    host.last_report = time if host.last_report.nil? or host.last_report.utc < time
```
What's last_report going to represent now?  This'll update it for all types of reports, but generally the dashboard behaviour is tied to the puppet_interval and outofsync_interval settings.

Should it just be on the config report importer?

domcleal commented on app/controllers/api/v1/reports_controller.rb:
```diff
@@ -11,7 +12,7 @@ class ReportsController < V1::BaseController
       def index
-        @reports = Report.
+        @reports = ConfigReport.
           authorized(:view_reports).
```
Lots of "view_reports" permissions in this PR, can you update them as the names changed?

domcleal replied:
Search the codebase and you'll also find DB seeds, fixtures and the dashboard helper need updating.

domcleal commented on test/unit/report_test.rb:
```diff
@@ -148,4 +148,28 @@ def setup
+
+    test "Report has child classes" do
+      my_report = MyReport.new
+      assert_equal(my_report.type, 'MyReport')
```
assertion's the wrong way around - expected, actual

shlomizadok replied:
Will change this for the next run :)

shlomizadok commented:
@domcleal, the `@report` instance variable and its `attr_reader` are very weird and I'm not sure about them.
In foreman, as well as in `ruby -Itest test/unit/report_test.rb` they work flawlessly. It fails for me only under `rake:test` with `NoMethodError: undefined method `persisted?' for nil:NilClass`.
I have pushed with that change (e.g., without setting `@report = super`).
I have also changed the menu text and globally changed the permissions from `:view_reports` to `view_config_reports`.

I have noticed your comment on report_test.rb, will push this after current tests are finished.

shlomizadok commented:
And I have moved `host.last_report` to config_report_importer

shlomizadok commented:
Report tests are failing: http://ci.theforeman.org/job/test_develop_pr_core/5973/,
Though it works of Foreman

domcleal commented on app/services/report_importer.rb:
```diff
@@ -108,10 +90,30 @@ def inspect_report
+  def create_report_and_logs
+    if host.new_record? && !Setting[:create_new_host_when_report_is_uploaded]
+      logger.info("skipping report for #{name} as its an unknown host and create_new_host_when_report_is_uploaded setting is disabled")
+      return report_name_class.new
```
`@report` is unset here, which is why the test is failing.

domcleal commented:
> In foreman, as well as in ruby -Itest test/unit/report_test.rb they work flawlessly. It fails for me only under rake:test with NoMethodError: undefined method `persisted?' for nil:NilClass.
> 
> Report tests are failing: http://ci.theforeman.org/job/test_develop_pr_core/5973/,

The test that's failing is a _host_ unit test (of report imports), not a report unit test.  I put an inline comment where I think it's probably failing, due to a short-circuit in the report importer code.

shlomizadok commented:
@domcleal Thank you so much!!! I was pulling my hear around this issue.
Short-circuit fixed

domcleal commented on app/models/report.rb:
```diff
@@ -122,8 +124,8 @@ def self.expire(conditions = {})
-    count = Report.where(cond).delete_all
-    logger.info Time.now.to_s + ": Expired #{count} Reports"
+    count = self.where(cond).delete_all
+    logger.info Time.now.to_s + ": Expired #{count} #{self.to_s.underscore.humanize.pluralize}"
```
`self` is superfluous on these two lines I think

domcleal commented on app/services/config_report_importer.rb:
```diff
@@ -0,0 +1,23 @@
+class ConfigReportImporter < ReportImporter
+  attr_reader :report
```
superfluous if it's in the parent?

domcleal commented on app/services/foreman/access_permissions.rb:
```diff
@@ -700,16 +700,19 @@
-    map.permission :upload_reports,  {:"api/v2/reports" => [:create] }
+  permission_set.security_block :config_reports do |map|
+    map.permission :view_config_reports, {:"api/v1/reports"        => [:index, :show, :last],
+                                   :"api/v2/reports"        => [:index, :show, :last],
```
reindent these lines to match the style, please

domcleal commented on db/migrate/20150728122736_change_report_permissions.rb:
```diff
@@ -0,0 +1,13 @@
+class ChangeReportPermissions < ActiveRecord::Migration
+  PERMISSIONS = %w(view_reports destroy_reports upload_reports)
+  def up
```
missing `down`?

domcleal commented on db/seeds.d/03-permissions.rb:
```diff
@@ -123,9 +123,9 @@
-  ['Report', 'view_reports'],
-  ['Report', 'destroy_reports'],
-  ['Report', 'upload_reports'],
+  ['ConfigReport', 'view_config_reports'],
```
This file's alphabetical, please fix the position

shlomizadok replied:
The whole file's abc ordering seems off. Fixing globally

domcleal commented on test/unit/report_importer_test.rb:
```diff
@@ -10,7 +10,7 @@ class ReportImporterTest < ActiveSupport::TestCase
 
   test 'it should import reports with no metrics' do
-    r = ReportImporter.import(read_json_fixture('report-empty.json'))
+    r = ConfigReportImporter.import(read_json_fixture('report-empty.json'))
```
Wrong class?

shlomizadok replied:
No, ReportImporter will raise NotImplementedError

domcleal commented on test/unit/config_report_importer_test.rb:
```diff
@@ -0,0 +1,57 @@
+      @host = FactoryGirl.create(:host, :owner => @owner)
+    end
+
+    test 'when owner is subscribed to notification, a mail should be sent on error' do
```
Shouldn't these be in the base class test, since that's where notifications are?

domcleal commented:
There's still quite a bit of support for metrics (applied, changed, failed_restarts etc) in Report that looks rather config management specific.  How will these be used in other subclasses?  Should they remain here or move - or would subclasses just use the fields that are most applicable to them?

shlomizadok commented:
@domcleal - I was expecting to something like:
merged as [23e5g7h](https://foreman-shlomizadok.rhcloud.com/2015/08/27/on-submitting-prs/) thanks @shlomizadok
However you've got valid comments / questions. Will attend to them later today / Sunday
Thanks!

shlomizadok commented:
regarding https://github.com/theforeman/foreman/pull/2530#issuecomment-135728473 - I am not sure.
I'd say the latter - subclasses use the field that are most applicable to them. What says you?

shlomizadok commented:
@domcleal - fixed the comments above, but https://github.com/theforeman/foreman/pull/2530#issuecomment-135728473 - please instruct me how to proceed there. Thanks

shlomizadok commented:
@domcleal - came to think of it and I'd like to keep the metrics in the current place so other subclasses may use them. If we will see that this is not the case and only config_report uses them we'll refactor on a separate pull request.

shlomizadok committed:
refs #11579 - Addition of PR #2644 - Reports show/destroy not restricted by host authorization

shlomizadok commented:
Manually added #2644. Refactored to reports STI. @eLobato, mind taking a look at the changes?

domcleal commented:
It needs rebasing to include that change, not added as a further commit.

shlomizadok commented:
@domcleal - rebased. thanks

shlomizadok commented:
tests :green_book: & rubocop :cop: are :green_heart:

shlomizadok commented:
Rebased

ares commented:
@shlomizadok  is this meant to be closed?

shlomizadok commented:
@ares yes
